### PR TITLE
Allow file: for requires statements in setup.cfg

### LINKED ISSUE #1951: Please allow "file:" for setup.cfg install_requires
The new declarative setup.cfg syntax is fabulous.  It's wonderful to have two-line setup.py files!

I tried to use "install_requires = file: requirements/base.in" and found that "file:" wasn't supported there.  Our pip-compile workflow works really well, and it would be great to be able to start with base.in as our source of truth, and easily use it in our setup.cfg.

As a design point, I'm not sure it makes sense to special-case which fields can use which kinds of info-gatherers.  Why can't I use "file:" for any field?  I can understand if the implementation requires it, but the more the developer can choose how to structure their own world, the better.

### PR DESCRIPTION
**Pull request by akx**
## Summary of changes

As requested and discussed in #1951, this PR enables using `file:` in <del>the various `requires` statements</del> `install_requires` and `extras_require` statements in setup.cfg.

As @jaraco mentioned in https://github.com/pypa/setuptools/issues/1951#issuecomment-574737840, this adds a (small) admonition advising against library packagers pinning their dependencies too tightly with this feature.

Closes #1951

### Pull Request Checklist

- [x] Changes have tests
- [x] Changes are documented 
- [x] News fragment added in `changelog.d/`.

### CODE REVIEW

**Review by abravalheri — COMMENTED** [review]
Hi @akx, thank you very much for analysing the previous discussion and providing the PR.

I will leave the review to the  maintainers that were participating in the original discussion, but I would like to note that this implementation does not seem to be *forward* compatible, because it only tackles `setup.cfg`.

According to https://github.com/pypa/setuptools/issues/1688#issuecomment-471212342, the long term view is to compile `setup.cfg` into `pyproject.toml`, therefore it would be necessary to support this feature in `pyproject.toml` as well.

**Comment by akx** [comment]
> [...] the long term view is to compile `setup.cfg` into `pyproject.toml`, therefore it would be necessary to support this feature in `pyproject.toml` as well.

Ah, sure thing! I can also take a look at getting that in here, if the approach seems otherwise sound by maintainers :)

**Comment by abravalheri** [comment]
Thank you very much for considering that @akx. Yes, it is definitely good to wait for the feedback before doing any change.

**Comment by jaraco** [comment]
I’m okay to give it a try.

**Comment by akx** [comment]
@abravalheri Hmm, by the looks of it enabling something like 
```
[tool.setuptools.dynamic.optional-dependencies.bonus]
file = ["requirements-bonus.txt"]
```
(along the lines of https://github.com/pypa/setuptools/blob/e5552d3c42eddc463788861ae4c8847183eeec1c/setuptools/tests/config/test_pyprojecttoml.py#L66-L71)

would require changes to the JSON Schema for pyproject.toml, which... apparently is maintained in `validate_pyproject`? Am I on the right track there?

If so, maybe all that should be done in a separate constellation of PRs..?

**Comment by abravalheri** [comment]
Hi @akx, If you change the code in setuptools, I am more than happy to follow up and tackle the schemas myself.

For the time being, in the tests you can use the `ignore_option_errors` parameter to bypass the validation. Once I modify the schemas, I will also modify that.

**Comment by akx** [comment]
@abravalheri Please see #3255 for the TOML case :)

File: docs/userguide/declarative_config.rst
Comment (abravalheri) [inline docs/userguide/declarative_config.rst:199]: Question: would it make more sense to restrict this change to `install_requires` and `extras_requires`?

`tests_require` is deprecated, so I feel like we should not touch it.

`setup_requires` is largely obsoleted by PEP 517.
In very specific use cases people might still reach for it inside `setup.py`, but not in the context of `setup.cfg`.

I understand that adding support for the `file:` directive for `setup_requires` and `tests_require` basically come for free, but I still think we should refrain from doing that.

Comment (abravalheri) [inline docs/userguide/declarative_config.rst:199]: Question: would it make more sense to restrict this change to `install_requires` and `extras_requires`?

`tests_require` is deprecated, so I feel like we should not touch it.

`setup_requires` is largely obsoleted by PEP 517.
In very specific use cases people might still reach for it inside `setup.py`, but not in the context of `setup.cfg`.

I understand that adding support for the `file:` directive for `setup_requires` and `tests_require` basically come for free, but I still think we should refrain from doing that.

Comment (abravalheri) [inline docs/userguide/declarative_config.rst:199]: Question: would it make more sense to restrict this change to `install_requires` and `extras_requires`?

`tests_require` is deprecated, so I feel like we should not touch it.

`setup_requires` is largely obsoleted by PEP 517.
In very specific use cases people might still reach for it inside `setup.py`, but not in the context of `setup.cfg`.

I understand that adding support for the `file:` directive for `setup_requires` and `tests_require` basically come for free, but I still think we should refrain from doing that.

Comment (akx) [reply docs/userguide/declarative_config.rst:199]: I get where you're coming from here, but various comments in #1951 say `file:` support is the last thing preventing them from moving from an imperative `setup.py` to a PEP 518 `setup.cfg` build. So, if those folks had been using files for their `setup_requires` and `tests_require`, not allowing `file:` for them would make that porting work less straightforward.

I have no other strong opinion on this, just that thought :)

Comment (abravalheri) [reply docs/userguide/declarative_config.rst:199]: Thank you very much @akx for pointing this out. I re-read the discussion in issue 1951 and I did not see any special request for `setup_requires` and `tests_requires`.

My view is that `tests_requires`, already does not work, so I don't think it would be helping anyone interested in the migration.

On the other hand, `setup_requires` only have value (right now) when computed dynamically via `setup.py`.

However I can see people starting to use it once it gets `file:` support just to workaround limitations in whichever GitHub bots they might be using for dependency management. In my opinion that would be an unfortunate side effect, since PEP 518 should be preferred.

---

On a side note, considering the forward compatibility with `pyproject.toml`:

- Currently `setup_requires` can be automatically converted to `[build-system] requires` by an automatic generator.
- If `setup_requires` gains the capability of being dynamically defined via `file:`, that will no longer be possible and we will have to update #3255 to handle this new use case.

**Comment by akx** [comment]
@abravalheri @jaraco I rebased this to get rid of the merge conflicts. What are the next steps to get this merged..?

**Comment by abravalheri** [comment]
Thank you very much @akx for rebasing the PR.

Personally, I have some strong reservations against handling dynamic `tests_require` and `setup_requires` (the main reasons are described in my previous comments https://github.com/pypa/setuptools/pull/3253#discussion_r846075228 and https://github.com/pypa/setuptools/pull/3253#discussion_r846242876).

Can we restrict the change to `install_requires` and `extras_require`?

**Comment by akx** [comment]
@abravalheri Okiedoke :) Removed the `file:` stuff related to `tests_require` and `setup_requires` (the code is still gently touched in `setupcfg.py` since `parse_list_semicolon` became a classmethod).

**Comment by abravalheri** [comment]
Thank you very much @akx and sorry for the trouble.
Since Jason seems to be OK with the change, I will try to merge the PRs yet this week.
(I am currently focusing on organising changes in the documentation and I had in the backlog the PEP 660, but I will try to make my best to not delay even more this merge).

**Review by abravalheri — APPROVED** [review]

**Comment by abravalheri** [comment]
Sorry for the delay @akx.

I merged the changes and published a new version of setuptools.
For the time being, I have added the status of *beta*, then we can see the feedback from the community.

**Comment by nedbat** [comment]
Thanks for this! :)

**Comment by pganssle** [comment]
I'm still pretty negative on this "feature".

**Comment by henryiii** [comment]
Just say my first usage of this in the wild: https://github.com/Flexget/Flexget/blob/f55a297fae18e9779d9df9e545974727ad39b07a/requirements.txt :)

**Pull request merged**
